components/core/src/clp_s/search/OutputHandler.hpp

```diff
@@ -0,0 +1,112 @@
+#ifndef CLP_S_SEARCH_OUTPUTHANDLER_HPP
+#define CLP_S_SEARCH_OUTPUTHANDLER_HPP
+
+#include <string>
+
+#include <mongocxx/client.hpp>
+#include <mongocxx/collection.hpp>
+#include <mongocxx/exception/exception.hpp>
+#include <mongocxx/instance.hpp>
+#include <mongocxx/uri.hpp>
+
+#include "../Defs.hpp"
+
+namespace clp_s::search {
+class OutputHandler {
+public:
+    explicit OutputHandler(bool output_timestamp, bool decompress_full_log = true)
+            : m_output_timestamp(output_timestamp) {}
+
+    virtual ~OutputHandler() = default;
+
+    virtual void write(std::string const& message, epochtime_t timestamp) = 0;
+    virtual void write(std::string const& message) = 0;
+    virtual void flush() = 0;
+
+    bool output_timestamp() const { return m_output_timestamp; }
+
+protected:
+    bool m_output_timestamp;
+};
+
+class StandardOutputHandler : public OutputHandler {
+public:
+    explicit StandardOutputHandler(bool output_timestamp = false)
+            : OutputHandler(output_timestamp) {}
+
+    void write(std::string const& message, epochtime_t timestamp) override {
+        printf("%lld %s", timestamp, message.c_str());
+    }
+
+    void flush() override {}
+
+    void write(std::string const& message) override { printf("%s", message.c_str()); }
+};
+
+class ResultsCacheOutputHandler : public OutputHandler {
+public:
+    class OperationFailed : public TraceableException {
+    public:
+        // Constructors
+        OperationFailed(ErrorCode error_code, char const* const filename, int line_number)
+                : TraceableException(error_code, filename, line_number) {}
+    };
+
+    ResultsCacheOutputHandler(
+            std::string const& uri,
+            std::string const& collection,
+            uint64_t batch_size,
+            bool output_timestamp = true
+    )
+            : OutputHandler(output_timestamp),
+              m_batch_size(batch_size) {
+        try {
+            auto mongo_uri = mongocxx::uri(uri);
+            m_client = mongocxx::client(mongo_uri);
+            m_collection = m_client[mongo_uri.database()][collection];
+        } catch (mongocxx::exception const& e) {
+            throw OperationFailed(ErrorCode::ErrorCodeBadParamDbUri, __FILENAME__, __LINE__);
+        }
+    }
+
+    void flush() override {
+        try {
+            if (!m_results.empty()) {
+                m_collection.insert_many(m_results);
+                m_results.clear();
+            }
+        } catch (mongocxx::exception const& e) {
+            throw OperationFailed(ErrorCode::ErrorCodeFailureDbBulkWrite, __FILENAME__, __LINE__);
+        }
+    }
+
+    void write(std::string const& message, epochtime_t timestamp) override {
+        try {
+            auto document = bsoncxx::builder::basic::make_document(
+                    bsoncxx::builder::basic::kvp("path", "logs.ndjson"),
```

Comment:
Yeah, but since clp and clp-s don't share the code and clp-s doesn't include a file for constant declarations, do we want to make them constants in their own classes?

Reply:
We could share this code (using a relative path), right?

Reply:
Yeah, but that makes the code a little bit awkward (because we haven't begun refactoring and merging two code bases). If we want to make them shared constants, do you think it's a good idea to put them in Defs.h right after `constexpr char cDefaultConfigFilename[] = ".clp.rc";` (in this way we need to include two Defs.h lol). Or we could declare it in one of the results cache file.

Reply:
Mmm, I sorta see what you're saying. I guess we can leave them separate for now. I would still make them consistent though (if we're keeping "path").